### PR TITLE
Update migraphx to rocm4.2

### PR DESCRIPTION
**Pull request by scxiao**
**Description**: Describe your changes.
   Fixed a build error related to update cmake to 20.3
   Added support for more operators
   upgraded to rocm4.2

**Motivation and Context**
  Make onnxruntime support up to date

- Why is this change required? What problem does it solve?
  Fixed build issues, more features are added.

- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: onnxruntime/core/providers/migraphx/migraphx_execution_provider.cc [hunk 1]
```diff
@@ -985,6 +1118,11 @@ Status MIGraphXExecutionProvider::Compile(const std::vector<onnxruntime::Node*>&
     std::string onnx_string_buffer;
     model_proto.SerializeToString(&onnx_string_buffer);
 
+    std::string name = "ort_compile_" + std::to_string(fused_node_index++) + ".onnx";
```
Comment (snnn) [inline onnxruntime/core/providers/migraphx/migraphx_execution_provider.cc:1121]: Could you please change std::ofstream to STD C APIs(fopen/fwrite/fclose) ? ofstream doesn't work well in our environment.

**Review by snnn — APPROVED** [review]
LGTM.

**Review by weixingzhang — APPROVED** [review]

**Pull request merged**